### DISCUSSION TITLE
Confusing wording in regards to ExecuteListeners: "In other words, listeners have the same lifetime as a single query execution"

### DISCUSSION THREAD

**Discussion by ghost** [jOOQ/jOOQ#15900] [Q&A] [answered]
Hello

I'm trying to implement thread-safe ExecuteListeners and stumbled upon this wording in the docs (link: https://www.jooq.org/javadoc/3.12.x/org.jooq/org/jooq/ExecuteListener.html)

"In other words, listeners have the same lifetime as a single query execution, and can thus be used to store state between the moment when a query execution starts, and the moment when a query execution finishes"

But if you were to use the DefaultExecuteListenerProvider and pass in a listener instance, the lifetime would actually be per DSLContext, right (or depending on how it's done, per Configuration)? I'm asking because we're storing state inside an ExecuteListener. I suppose the solution here would be to use the CallbackExecuteListener or create a custom one? Sort of a PerQueryExecuteListenerProvider? I looked at the ExecuteListeners and AbstractQuery code for examples.

**Comment by lukaseder** [comment, marked as answer]
Thanks for your comment. That wording is wrong (or probably very outdated, I vaguely recall execute listeners being instantiated using reflection in the distant past).

I'll fix it right away:

- https://github.com/jOOQ/jOOQ/issues/15901

The `ExecuteListenerProvider` is really the instance that governs `ExecuteListener` lifecycles, and you're in full control of that. jOOQ will call `ExecuteListenerProvider::provide`.

You can also store data in `ExecuteContext::data`, by the way.

**Reply by ghost** [reply]
I didn't know the thing about ExecuteContext::data.

Thank you!
